README.md
# Altair
Altair is a UCI-protocol chess engine that succeeds my old python engine Antares.

## Rating
CCRL Ratings:

| Version | Blitz | Rapid |
|:-------:|:-----:|:-----:|
|  1.0.0  | ----  | 2409  |


## Download
Go to releases for binaries. You can build from the source code with cmake.

## Play
You can play Altair [here](https://lichess.org/@/Altair_Engine)
It is not always online.

Altair can also be played by connecting it to GUIs such as Arena, Cutechess, Banksia, and other UCI-protocol supporting 
chess GUIs.

## Board Representation
Altair uses a 10x12 mailbox array. This board representation gives padding around the 8x8 board for faster move generation. Altair also features
piece lists.

I hope to use bitboards instead one day since they can be much faster if implemented well.

## Move Generation
Offsets / Increments are used in loops to generate the pseudo-legal moves. Within the search and perft, the moves are tested for 
legality, and if the move is not legal, we skip to the next move.

## Search
#### Iterative Deepening
- Aspiration Windows
  - Negamax (Minimax)
    - Alpha-Beta Pruning
      - Reverse Futility Pruning
      - Null Move Pruning
      - Idea similar to Internal Iterative Deepening
      - Principal Variation Search
      - Late Move Reductions
    - Transposition Tables
  - Quiescence Search
    - Delta Pruning
    - Transposition Tables
  
#### Move Ordering
- Transposition Table Move
- Captures
  - Most Valuable Victim - Least Valuable Agressor (MVV-LVA)
  - En passant
- Killer move Heuristic
- Counter move Heuristic
- History move Heuristic
- Promotions
- Castling

## Evaluation
- Material
- Piece Square Tables
- Mobility
- Double, Isolated, Backwards, and Passed Pawns
- Open and semi-open files for rooks, queens and kings
- King's pawn-shield
- Bishop Pair
- Tempo
- King Tropism (Distance from pieces to a King)
- Tapered evaluation between the middlegame and endgame

## Credit and Thanks
- [The Chess Programming Wiki](https://www.chessprogramming.org/Main_Page) contains invaluable resources that I have used.
- [Bitboard Chess Engine in C](https://www.youtube.com/playlist?list=PLmN0neTso3Jxh8ZIylk74JpwfiWNI76Cs) A nice series that contains implementations
for different ideas and features.
- Gedas and his engine [Chessdotcpp](https://github.com/GediminasMasaitis/chess-dot-cpp) for different ideas and implementations.
- Gedas' [Tuner](https://github.com/GediminasMasaitis/texel-tuner/tree/main/src) for tuning evaluation parameters
- [Weiss](https://github.com/TerjeKir/weiss) Another good resource.
- Engine Programming Discord Server for answering many of my questions


changelog
# Changelog

### Bare a/b negamax with PST eval
```
info depth 1 score cp 58 time 0 nodes 21 nps 210000 pv b1c3
info depth 2 score cp 0 time 1 nodes 127 nps 127000 pv b1c3 g8f6
info depth 3 score cp 58 time 3 nodes 1246 nps 415333 pv b1c3 g8f6 g1f3
info depth 4 score cp 0 time 6 nodes 8680 nps 1446666 pv b1c3 b8c6 g1f3 g8f6
info depth 5 score cp 55 time 41 nodes 184415 nps 4497926 pv e2e4 b8c6 f1b5 c6e5 b5d7
info depth 6 score cp -30 time 274 nodes 1326069 nps 4839667 pv b1c3 e7e5 e2e3 f8b4 c3d5 b4d2
info depth 7 score cp 98 time 3277 nodes 17673132 nps 5393082 pv e2e3 b8c6 b1c3 c6b4 f1c4 g8f6 c4f7
info depth 8 score cp -63 time 41540 nodes 213790795 nps 5146624 pv b1c3 b8c6 e2e3 e7e5 f1b5 f8b4 c3d5 b4d2
info depth 8 score cp -63 time 44999 nodes 231026698 nps 5134040 pv b1c3 b8c6 e2e3 e7e5 f1b5 f8b4 c3d5 b4d2
bestmove b1c3
```


### Qsearch + pre-allocating move vectors
```
info depth 1 score cp 58 time 0 nodes 21 nps 210000 pv b1c3
info depth 2 score cp 0 time 2 nodes 157 nps 78500 pv b1c3 b8c6
info depth 3 score cp 58 time 4 nodes 2362 nps 590500 pv b1c3 g8f6 g1f3
info depth 4 score cp 0 time 7 nodes 14721 nps 2103000 pv g1f3 g8f6 b1c3 b8c6
info depth 5 score cp 41 time 35 nodes 157226 nps 4492171 pv b1c3 g8f6 g1f3 b8c6 d2d4
info depth 6 score cp 0 time 412 nodes 1765136 nps 4284310 pv b1c3 b8c6 g1f3 g8f6 e2e4 d7d5
info depth 7 score cp 30 time 7861 nodes 31894942 nps 4057364 pv e2e4 b8c6 g1f3 g8f6 e4e5 f6e4 b1c3
info depth 7 score cp 30 time 45000 nodes 146612255 nps 3258050 pv e2e4 b8c6 g1f3 g8f6 e4e5 f6e4 b1c3
bestmove e2e4

Score of AntaresCpp0.01 vs AntaresCpp0.00: 54 - 5 - 41 [0.745]
...      AntaresCpp0.01 playing White: 23 - 1 - 26  [0.720] 50
...      AntaresCpp0.01 playing Black: 31 - 4 - 15  [0.770] 50
...      White vs Black: 27 - 32 - 41  [0.475] 100
Elo difference: 186.2 +/- 53.7, LOS: 100.0 %, DrawRatio: 41.0 %

Score of AntaresCpp0.01 vs SebLagueEngine0.1: 6 - 18 - 5 [0.293]
...      AntaresCpp0.01 playing White: 3 - 9 - 3  [0.300] 15
...      AntaresCpp0.01 playing Black: 3 - 9 - 2  [0.286] 14
...      White vs Black: 12 - 12 - 5  [0.500] 29
Elo difference: -152.9 +/- 132.7, LOS: 0.7 %, DrawRatio: 17.2 %
```


### In check extensions
```
info depth 1 score cp 58 time 0 nodes 21 nps 210000 pv b1c3
info depth 2 score cp 0 time 1 nodes 157 nps 157000 pv b1c3 b8c6
info depth 3 score cp 58 time 2 nodes 2362 nps 1181000 pv b1c3 g8f6 g1f3
info depth 4 score cp 0 time 5 nodes 14756 nps 2951200 pv g1f3 g8f6 b1c3 b8c6
info depth 5 score cp 41 time 32 nodes 150351 nps 4698468 pv b1c3 g8f6 g1f3 b8c6 d2d4
info depth 6 score cp 0 time 510 nodes 2111693 nps 4140574 pv d2d4 d7d5 g1f3 g8f6 b1c3 b8c6
info depth 7 score cp 30 time 4875 nodes 21066927 nps 4321420 pv e2e4 b8c6 g1f3 g8f6 e4e5 f6e4 b1c3
info depth 7 score cp 30 time 44999 nodes 171016216 nps 3800444 pv e2e4 b8c6 g1f3 g8f6 e4e5 f6e4 b1c3
bestmove e2e4

Score of AntaresCpp0.02 vs AntaresCpp0.01: 21 - 20 - 59 [0.505]
...      AntaresCpp0.02 playing White: 16 - 5 - 29  [0.610] 50
...      AntaresCpp0.02 playing Black: 5 - 15 - 30  [0.400] 50
...      White vs Black: 31 - 10 - 59  [0.605] 100
Elo difference: 3.5 +/- 43.8, LOS: 56.2 %, DrawRatio: 59.0 %

Score of AntaresCpp0.02 vs SebLagueEngine0.1: 15 - 103 - 32 [0.207]
...      AntaresCpp0.02 playing White: 7 - 50 - 18  [0.213] 75
...      AntaresCpp0.02 playing Black: 8 - 53 - 14  [0.200] 75
...      White vs Black: 60 - 58 - 32  [0.507] 150
Elo difference: -233.7 +/- 57.4, LOS: 0.0 %, DrawRatio: 21.3 %
```


### Negamax Move Ordering
```
info depth 1 score cp 58 time 0 nodes 21 nps 210000 pv b1c3
info depth 2 score cp 0 time 1 nodes 157 nps 157000 pv b1c3 b8c6
info depth 3 score cp 58 time 3 nodes 2176 nps 725333 pv b1c3 b8c6 g1f3
info depth 4 score cp 0 time 7 nodes 13851 nps 1978714 pv g1f3 b8c6 b1c3 g8f6
info depth 5 score cp 41 time 24 nodes 91403 nps 3808458 pv g1f3 b8c6 d2d4 g8f6 b1c3
info depth 6 score cp 0 time 267 nodes 1000631 nps 3747681 pv g1f3 g8f6 d2d4 d7d5 b1c3 b8c6
info depth 7 score cp 30 time 2067 nodes 7725204 nps 3737399 pv e2e4 b8c6 g1f3 g8f6 e4e5 f6e4 b1c3
info depth 8 score cp 0 time 17696 nodes 61743103 nps 3489099 pv g1f3 g8f6 e2e3 b8c6 b1c3 d7d6 d2d4 e7e5
info depth 8 score cp 0 time 45000 nodes 156903437 nps 3486743 pv g1f3 g8f6 e2e3 b8c6 b1c3 d7d6 d2d4 e7e5
bestmove g1f3

Score of AntaresCpp0.03 vs AntaresCpp0.02: 80 - 9 - 62 [0.735]
...      AntaresCpp0.03 playing White: 45 - 3 - 28  [0.776] 76
...      AntaresCpp0.03 playing Black: 35 - 6 - 34  [0.693] 75
...      White vs Black: 51 - 38 - 62  [0.543] 151
Elo difference: 177.3 +/- 43.5, LOS: 100.0 %, DrawRatio: 41.1 %

Score of AntaresCpp0.03 vs SebLagueEngine0.1: 34 - 30 - 36 [0.520]
...      AntaresCpp0.03 playing White: 18 - 13 - 19  [0.550] 50
...      AntaresCpp0.03 playing Black: 16 - 17 - 17  [0.490] 50
...      White vs Black: 35 - 29 - 36  [0.530] 100
Elo difference: 13.9 +/- 54.9, LOS: 69.1 %, DrawRatio: 36.0 %
```


### Quiescence Search Move Ordering
```
info depth 1 score cp 58 time 0 nodes 21 nps 210000 pv b1c3
info depth 2 score cp 0 time 1 nodes 157 nps 157000 pv b1c3 b8c6
info depth 3 score cp 58 time 2 nodes 2176 nps 1088000 pv b1c3 b8c6 g1f3
info depth 4 score cp 0 time 6 nodes 15135 nps 2522500 pv g1f3 g8f6 b1c3 b8c6
info depth 5 score cp 41 time 20 nodes 78230 nps 3911500 pv g1f3 b8c6 d2d4 g8f6 b1c3
info depth 6 score cp 0 time 167 nodes 607033 nps 3634928 pv g1f3 d7d5 b1c3 b8c6 e2e4 g8f6
info depth 7 score cp 30 time 2349 nodes 9543601 nps 4062835 pv e2e4 b8c6 g1f3 g8f6 e4e5 f6e4 b1c3
info depth 8 score cp 0 time 19282 nodes 71098146 nps 3687280 pv d2d4 g8f6 e2e3 d7d5 b1c3 b8c6 g1f3 c8f5
info depth 8 score cp 0 time 44999 nodes 152193051 nps 3382142 pv d2d4 g8f6 e2e3 d7d5 b1c3 b8c6 g1f3 c8f5
bestmove d2d4

Score of AntaresCpp0.04 vs AntaresCpp0.03: 60 - 35 - 105 [0.563]
...      AntaresCpp0.04 playing White: 32 - 16 - 52  [0.580] 100
...      AntaresCpp0.04 playing Black: 28 - 19 - 53  [0.545] 100
...      White vs Black: 51 - 44 - 105  [0.517] 200

bench.h
//
// Created by Alexander Tian on 4/12/23.
//

#ifndef ALTAIRCHESSENGINE_BENCH_H
#define ALTAIRCHESSENGINE_BENCH_H

#include "constants.h"
#include "search.h"
#include "position.h"

constexpr std::array fens { // fens stolen from polaris, which were from alexandria, and ultimately from bit-genie
        "r3k2r/2pb1ppp/2pp1q2/p7/1nP1B3/1P2P3/P2N1PPP/R2QK2R w KQkq a6 0 14",
        "4rrk1/2p1b1p1/p1p3q1/4p3/2P2n1p/1P1NR2P/PB3PP1/3R1QK1 b - - 2 24",
        "r3qbrk/6p1/2b2pPp/p3pP1Q/PpPpP2P/3P1B2/2PB3K/R5R1 w - - 16 42",
        "6k1/1R3p2/6p1/2Bp3p/3P2q1/P7/1P2rQ1K/5R2 b - - 4 44",
        "8/8/1p2k1p1/3p3p/1p1P1P1P/1P2PK2/8/8 w - - 3 54",
        "7r/2p3k1/1p1p1qp1/1P1Bp3/p1P2r1P/P7/4R3/Q4RK1 w - - 0 36",
        "r1bq1rk1/pp2b1pp/n1pp1n2/3P1p2/2P1p3/2N1P2N/PP2BPPP/R1BQ1RK1 b - - 2 10",
        "3r3k/2r4p/1p1b3q/p4P2/P2Pp3/1B2P3/3BQ1RP/6K1 w - - 3 87",
        "2r4r/1p4k1/1Pnp4/3Qb1pq/8/4BpPp/5P2/2RR1BK1 w - - 0 42",
        "4q1bk/6b1/7p/p1p4p/PNPpP2P/KN4P1/3Q4/4R3 b - - 0 37",
        "2q3r1/1r2pk2/pp3pp1/2pP3p/P1Pb1BbP/1P4Q1/R3NPP1/4R1K1 w - - 2 34",
        "1r2r2k/1b4q1/pp5p/2pPp1p1/P3Pn2/1P1B1Q1P/2R3P1/4BR1K b - - 1 37",
        "r3kbbr/pp1n1p1P/3ppnp1/q5N1/1P1pP3/P1N1B3/2P1QP2/R3KB1R b KQkq b3 0 17",
        "8/6pk/2b1Rp2/3r4/1R1B2PP/P5K1/8/2r5 b - - 16 42",
        "1r4k1/4ppb1/2n1b1qp/pB4p1/1n1BP1P1/7P/2PNQPK1/3RN3 w - - 8 29",
        "8/p2B4/PkP5/4p1pK/4Pb1p/5P2/8/8 w - - 29 68",
        "3r4/ppq1ppkp/4bnp1/2pN4/2P1P3/1P4P1/PQ3PBP/R4K2 b - - 2 20",
        "5rr1/4n2k/4q2P/P1P2n2/3B1p2/4pP2/2N1P3/1RR1K2Q w - - 1 49",
        "1r5k/2pq2p1/3p3p/p1pP4/4QP2/PP1R3P/6PK/8 w - - 1 51",
        "q5k1/5ppp/1r3bn1/1B6/P1N2P2/BQ2P1P1/5K1P/8 b - - 2 34",
        "r1b2k1r/5n2/p4q2/1ppn1Pp1/3pp1p1/NP2P3/P1PPBK2/1RQN2R1 w - - 0 22",
        "r1bqk2r/pppp1ppp/5n2/4b3/4P3/P1N5/1PP2PPP/R1BQKB1R w KQkq - 0 5",
        "r1bqr1k1/pp1p1ppp/2p5/8/3N1Q2/P2BB3/1PP2PPP/R3K2n b Q - 1 12",
        "r1bq2k1/p4r1p/1pp2pp1/3p4/1P1B3Q/P2B1N2/2P3PP/4R1K1 b - - 2 19",
        "r4qk1/6r1/1p4p1/2ppBbN1/1p5Q/P7/2P3PP/5RK1 w - - 2 25",
        "r7/6k1/1p6/2pp1p2/7Q/8/p1P2K1P/8 w - - 0 32",
        "r3k2r/ppp1pp1p/2nqb1pn/3p4/4P3/2PP4/PP1NBPPP/R2QK1NR w KQkq - 1 5",
        "3r1rk1/1pp1pn1p/p1n1q1p1/3p4/Q3P3/2P5/PP1NBPPP/4RRK1 w - - 0 12",
        "5rk1/1pp1pn1p/p3Brp1/8/1n6/5N2/PP3PPP/2R2RK1 w - - 2 20",
        "8/1p2pk1p/p1p1r1p1/3n4/8/5R2/PP3PPP/4R1K1 b - - 3 27",
        "8/4pk2/1p1r2p1/p1p4p/Pn5P/3R4/1P3PP1/4RK2 w - - 1 33",
        "8/5k2/1pnrp1p1/p1p4p/P6P/4R1PK/1P3P2/4R3 b - - 1 38",
        "8/8/1p1kp1p1/p1pr1n1p/P6P/1R4P1/1P3PK1/1R6 b - - 15 45",
        "8/8/1p1k2p1/p1prp2p/P2n3P/6P1/1P1R1PK1/4R3 b - - 5 49",
        "8/8/1p4p1/p1p2k1p/P2npP1P/4K1P1/1P6/3R4 w - - 6 54",
        "8/8/1p4p1/p1p2k1p/P2n1P1P/4K1P1/1P6/6R1 b - - 6 59",
        "8/5k2/1p4p1/p1pK3p/P2n1P1P/6P1/1P6/4R3 b - - 14 63",
        "8/1R6/1p1K1kp1/p6p/P1p2P1P/6P1/1Pn5/8 w - - 0 67",
        "1rb1rn1k/p3q1bp/2p3p1/2p1p3/2P1P2N/PP1RQNP1/1B3P2/4R1K1 b - - 4 23",
        "4rrk1/pp1n1pp1/q5p1/P1pP4/2n3P1/7P/1P3PB1/R1BQ1RK1 w - - 3 22",
        "r2qr1k1/pb1nbppp/1pn1p3/2ppP3/3P4/2PB1NN1/PP3PPP/R1BQR1K1 w - - 4 12",
        "2r2k2/8/4P1R1/1p6/8/P4K1N/7b/2B5 b - - 0 55",
        "6k1/5pp1/8/2bKP2P/2P5/p4PNb/B7/8 b - - 1 44",
        "2rqr1k1/1p3p1p/p2p2p1/P1nPb3/2B1P3/5P2/1PQ2NPP/R1R4K w - - 3 25",
        "r1b2rk1/p1q1ppbp/6p1/2Q5/8/4BP2/PPP3PP/2KR1B1R b - - 2 14",
        "6r1/5k2/p1b1r2p/1pB1p1p1/1Pp3PP/2P1R1K1/2P2P2/3R4 w - - 1 36",
        "rnbqkb1r/pppppppp/5n2/8/2PP4/8/PP2PPPP/RNBQKBNR b KQkq c3 0 2",
        "2rr2k1/1p4bp/p1q1p1p1/4Pp1n/2PB4/1PN3P1/P3Q2P/2RR2K1 w - f6 0 20",
        "3br1k1/p1pn3p/1p3n2/5pNq/2P1p3/1PN3PP/P2Q1PB1/4R1K1 w - - 0 23",
        "2r2b2/5p2/5k2/p1r1pP2/P2pB3/1P3P2/K1P3R1/7R w - - 23 93"
};

void run_bench(Engine& engine, Position& position, PLY_TYPE depth);

#endif //ALTAIRCHESSENGINE_BENCH_H
